src/App.js

```diff
@@ -37,77 +128,429 @@ function App(props) {
+        updateDoc(doc(db, collectionName, userDocumentName), {
+            lists: newLists
+        });
+
+        // Update state to be on newly added list
+        setSubcollectionName(name);
     }
 
-    // Delete item by id
-    const onDeleteItem = (id) => {
-        setData(data.filter((item) => item.id !== id));
+    // Edit list name
+    const onEditList = (oldName, newName) => {
+        setUndoStack(undoStack.concat([{
+            type: UndoType.List,
+            op: UndoOp.Edit,
+            oldList: oldName,
+            newList: newName,
+            data: itemsData,
+        }]));
+
+        // Firestore does not support renaming collections
+        // To rename list, must delete old list and re-add all of its items under new collection name
+        // Use batch to make this operation atomic
+        const batch = writeBatch(db);
+
+        // Remove list from user lists array and replace with new list containing new list name
+        const newLists = userData.lists.map((list) => {
+            if (list === oldName) {
+                return newName;
+            } else {
+                return list
+            }
+        });
+        batch.update(doc(db, collectionName, userDocumentName), {
+            lists: newLists
+        });
+
+        // Delete all items under old list name
+        const oldItems = [...itemsData];
+        oldItems.forEach((item) => {
+            batch.delete(doc(db, collectionPath + subcollectionName, item.id));
+        })
+
+        // Add back all items under new list name
+        oldItems.forEach((item) => {
+            batch.set(doc(db, collectionPath + newName, item.id), {
+               id: item.id,
+               text: item.text,
+               completed: item.completed,
+               priority: item.priority,
+               created: item.created,
+            });
+        });
+
+        batch.commit();
+
+        // Update state to be on newly edited list
+        setSubcollectionName(newName);
     }
 
-    // Edit item
-    const onEditItem = (id, text, completed) => {
-        setData(data.map((item) => {
-            if (item.id === id) {
-                const newItem = {...item};
-                newItem.text = text;
-                newItem.completed = completed;
-                return newItem;
+    // Delete list
+    const onDeleteList = (name) => {
+        setUndoStack(undoStack.concat([{
+            type: UndoType.List,
+            op: UndoOp.Delete,
+            oldList: name,
+            newList: null,
+            data: itemsData,
+        }]));
+
+        // First delete all items in list
+        const batch = writeBatch(db);
+        itemsData.forEach((item) => {
+            batch.delete(doc(db, collectionPath + subcollectionName, item.id));
+        });
+        batch.commit();
+
+        // Then delete list itself from user array so it doesn't show in dropdown anymore
+        const newLists = userData.lists.filter(list => list !== subcollectionName);
```

Comment:
Is this necessary? Won't the lists be updated when useCollectionData returns the new lists?

Reply:
useCollectionData grabs the array of lists from the user document, so we have to remove the list from that array.